### DISCUSSION TITLE
How to use Pub/Sub for sentinel events

### DISCUSSION THREAD

**Discussion by Renboo** [redis/lettuce#3004] [Q&A] [answered]
Hello!

Please give some information about how to use how to use Pub/Sub for sentinel events.

I use Spring data with lettuce.
Connection to sentinel works good but by default lettuce do not accept any sentinel events.
Please give me some example how to configure Pub/Sub for sentinel events.

**Comment by Renboo** [comment]
I find in other disscussion some ideas:

1. Sub for sentinel events will work automatically
2. Add 
`builder.readFrom(ReadFrom.MASTER)`
in bean that return LettuceClientConfigurationBuilderCustomizer

Problems:
1. 1 - does not work 
2. 2 - result is error: 
`NOPERM User default has no permission to run the 'sentinel|replicas' command`

**Comment by tishun** [comment]
Hey @Renboo ,

Can you give some more details on what you want to achieve and also some code samples?

I think you might be confusing two different features here:
* one is [Redis Pub/Sub](https://redis.io/docs/latest/develop/interact/pubsub/) and it allows for applications to subscribe and publish to messaging channels (or topics), while
* the other is [Redis Sentinel](https://redis.io/docs/latest/operate/oss_and_stack/management/sentinel/) and it helps manage a clustered Redis deployment, by helping achieve a measure of High Availability

They are different features and can be used both together and independently from each other.

For more information on how to use them in Lettuce you can check out [Pub/sub in the Lettuce driver](https://redis.github.io/lettuce/#pubsub) and [Redis Sentinel in the Lettuce driver](https://redis.github.io/lettuce/ha-sharding/#redis-sentinel_2)

**Comment by Renboo** [comment]
Hey @tishun 

I have a requirement to use Sentinel with High Availability way. For me it means that first of all i get master node info from Sentinels and then i will subscribe/psubscribe to master changes events to change connection address as fast as possible.
Does lettuce uses that scenario when connected to Sentinels? Should i configure something more to execute my scenario? Or should all work out of the box?

**Reply by tishun** [reply, marked as answer]
> I have a requirement to use Sentinel with High Availability way. 

In order to use Sentinel toy do not need to use pub/sub, the driver would handle changes to the environment for you.
For more details see [Redis Sentinel](https://redis.github.io/lettuce/ha-sharding/#redis-sentinel_2) and [Redis discovery using Redis Sentinel](https://redis.github.io/lettuce/ha-sharding/#redis-discovery-using-redis-sentinel), specifically:

> The disconnect can result in any of the following options:
> 
> The master comes back: The connection is restored to the Redis instance
> 
> A replica is promoted to a master: Lettuce performs an address lookup using the masterId. As soon as the Redis Sentinel provides an address the connection is restored to the new Redis instance

**Reply by tishun** [reply]
Do you need any more help on this?

**Reply by Renboo** [reply]
@tishun Thank you for your reply!
It should work. I had some problems with user privileges what could be the reason of my problem.

**Reply by tishun** [reply]
Awesome, thanks for letting me know!
